superset/migrations/versions/fe23025b9441_rename_big_viz_total_form_data_fields.py

```diff
@@ -0,0 +1,92 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""rename_big_viz_total_form_data_fields
+
+Revision ID: fe23025b9441
+Revises: abe27eaf93db
+Create Date: 2021-12-13 14:06:24.426970
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = "fe23025b9441"
+down_revision = "abe27eaf93db"
+
+import json
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy import Column, Integer, String, Text
+from sqlalchemy.ext.declarative import declarative_base
+
+from superset import db
+
+Base = declarative_base()
+
+
+class Slice(Base):
+    __tablename__ = "slices"
+
+    id = Column(Integer, primary_key=True)
+    params = Column(Text)
+    viz_type = Column(String(250))
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    slices = session.query(Slice).filter(Slice.viz_type == "big_number_total").all()
+    for slc in slices:
+        try:
+            params = json.loads(slc.params)
+            header_format_selector = params.pop("header_format_selector", None)
+            header_timestamp_format = params.pop("header_timestamp_format", None)
+            if header_format_selector:
+                params["force_timestamp_formatting"] = header_format_selector
+            if header_timestamp_format:
+                params["time_format"] = header_timestamp_format
+            slc.params = json.dumps(params, sort_keys=True)
+        except Exception as e:
+            print(e)
+            print(f"Parsing params for slice {slc.id} failed.")
+
+    session.commit()
+    session.close()
+
+
+def downgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    slices = session.query(Slice).filter(Slice.viz_type == "big_number_total").all()
+    for slc in slices:
+        try:
+            params = json.loads(slc.params)
+            time_format = params.pop("time_format", None)
+            force_timestamp_formatting = params.pop("force_timestamp_formatting", None)
+            if time_format:
+                params["header_timestamp_format"] = time_format
+            if force_timestamp_formatting:
+                params["header_format_selector"] = force_timestamp_formatting
+            slc.params = json.dumps(params, sort_keys=True)
+        except Exception as e:
+            print(e)
+            print(f"Parsing params for slice {slc.id} failed.")
```

Comment:
I've seen a couple migrations that succeeded even when exceptions were raised, and they always had to be fixed later to address inconsistent states.

Here, it would be better to raise the exception, so that the migration fails and the admin has the chance to fix the offending charts before trying again. Otherwise, if the migration succeeds they might not even see the message, leaving their system inconsistent, and even if they see the message they need to downgrade and upgrade, which is more work.

Also, can you update your summary confirming that migrations are atomic, have been tested, and share the benchmark numbers? You can use `scripts/benchmark_migration.py` to get the numbers.

Reply:
Thanks for the comment! I updated the description and and re-raised the exception